File: Model/Source/Well.vb
﻿Imports System.ComponentModel
Imports Wells.Model

Public Class Well
    Inherits BusinessObject
    Implements IComparable(Of Well)

    <DisplayName("Nombre"), Browsable(True)>
    Property Name As String

    <DisplayName("X"), Browsable(True)>
    Property X As Double

    <DisplayName("Y"), Browsable(True)>
    Property Y As Double

    <DisplayName("Cota"), Browsable(True)>
    Property Z As Double

    <DisplayName("Latitud"), Browsable(True)>
    Property Latitude As Double

    <DisplayName("Longitud"), Browsable(True)>
    Property Longitude As Double

    <Browsable(False)>
    ReadOnly Property HasGeographic As Boolean
        Get
            Return Not (Latitude = NullNumericValue AndAlso Longitude = NullNumericValue)
        End Get
    End Property

    <Browsable(False)>
    ReadOnly Property HasHeight As Boolean
        Get
            Return Height <> NullNumericValue
        End Get
    End Property

    <Browsable(False)>
    Property Type As WellType

    <DisplayName("Altura"), Browsable(True)>
    Property Height As Double

    <DisplayName("Existe"), Browsable(True)>
    Property Exists As Boolean

    <DisplayName("Fondo"), Browsable(True)>
    Property Bottom As Double

    <Browsable(False)>
    Property Analysis As New List(Of ChemicalAnalysis)

    <Browsable(False)>
    Property Measurements As New List(Of Measurement)

    <Browsable(False)>
    Property Links As New List(Of ExternalLink)

    <Browsable(False)>
    ReadOnly Property SoilAnalysis As List(Of ChemicalAnalysis)
        Get
            Dim list = (From a In Analysis
                        Where a.SampleOf = SampleType.Soil
                        Select a).ToList
            Return list
        End Get
    End Property

    <Browsable(False)>
    ReadOnly Property WaterAnalysis As List(Of ChemicalAnalysis)
        Get
            Dim list = (From a In Analysis
                        Where a.SampleOf = SampleType.Water
                        Select a).ToList
            Return list
        End Get
    End Property

    <Browsable(False)>
    ReadOnly Property FLNAAnalysis As List(Of ChemicalAnalysis)
        Get
            Dim list = (From a In Analysis
                        Where a.SampleOf = SampleType.FLNA
                        Select a).ToList
            Return list
        End Get
    End Property

    Sub New()
        MyBase.New()
    End Sub

    Sub New(name As String)
        MyBase.New()
        Me.Name = name
    End Sub

    Public Overrides Function ToString() As String
        Return Name
    End Function

    Public Function CompareTo(other As Well) As Integer Implements IComparable(Of Well).CompareTo
        If Name > other.Name Then Return -1
        If Name = other.Name Then Return 0
        Return 1
    End Function
End Class

Public Enum WellType
    MeasurementWell
    Sounding
End Enum

File: Model/Model/src/Well.cs
﻿using System.Collections.Generic;
using System.ComponentModel;
using System.Reflection;
using Wells.BaseModel.Models;

namespace Wells.Model
{
    public class Well : BusinessObject
    {
        public Well()
        {
            SoilAnalyses = new List<SoilAnalysis>();
            WaterAnalyses = new List<WaterAnalysis>();
            FlnaAnalyses = new List<FlnaAnalysis>();
            Measurements = new List<Measurement>();
            Files = new List<ExternalFile>();
        }

        [Browsable(true), DisplayName("Nombre")]
        public string Name { get; set; }

        #region Properties
        [DisplayName("X"), Browsable(true)]
        public double X { get; set; }

        [DisplayName("Y"), Browsable(true)]
        public double Y { get; set; }

        [DisplayName("Cota (m.s.n.m.)"), Browsable(true)]
        public double Z { get; set; }

        [DisplayName("Latitud"), Browsable(true)]
        public double Latitude { get; set; }

        [DisplayName("Longitud"), Browsable(true)]
        public double Longitude { get; set; }

        [DisplayName("Altura del brocal (m)"), Browsable(true)]
        public double Height { get; set; }

        [DisplayName("Fondo (m.b.b.p.)"), Browsable(true)]
        public double Bottom { get; set; }

        [Browsable(false)]
        public bool HasGeographic => !(Latitude.Equals(NumericNullValue) || Longitude.Equals(NumericNullValue));

        [Browsable(false)]
        public bool HasHeight => !Height.Equals(NumericNullValue);

        [Browsable(false)]
        public bool HasZ => !Z.Equals(NumericNullValue);

        [DisplayName("Tipo"), Browsable(true)]
        public WellType WellType { get; set; }


        [DisplayName("Existe"), Browsable(true)]
        public bool Exists { get; set; }

        #endregion


        #region Lazy-loaded properties
        [Browsable(false)]
        public virtual List<SoilAnalysis> SoilAnalyses { get; set; }

        [Browsable(false)]
        public virtual List<WaterAnalysis> WaterAnalyses { get; set; }

        [Browsable(false)]
        public virtual List<FlnaAnalysis> FlnaAnalyses { get; set; }

        [Browsable(false)]
        public virtual List<Measurement> Measurements { get; set; }

        [Browsable(false)]
        public virtual List<ExternalFile> Files { get; set; }
        #endregion



        [Browsable(false)]
        public static Dictionary<string, PropertyInfo> Properties
        {
            get
            {
                return GetBrowsableProperties(typeof(Well));
            }
        }

        [Browsable(false)]
        public static Dictionary<string, PropertyInfo> DoubleProperties
        {
            get
            {
                return GetDoubleBrowsableProperties(typeof(Well));
            }
        }

        public override int CompareTo(IBusinessObject other)
        {
            return Name.CompareTo((other as Well).Name);
        }

        public static string GetDisplayName(string propertyName)
        {
            return GetDisplayName(typeof(Well), propertyName);
        }

        public override string ToString()
        {
            return Name;
        }
    }
}